084-services/proposal.md

```diff
@@ -0,0 +1,153 @@
+
+If `startup_probe.run` is left unspecified, Concourse will wait for each of the specified ports to be open.
+
+## Worker Placement
+
+Since `services` are just bound to `task`s, the easiest approach would be to assign the service container and the task container to the same worker. This allows us to avoid a more complex architecture having to route traffic through the TSA (since workers may not be directly reachable from one another).
+
+This hopefully isn't *too* restrictive, as anyone running e.g. `docker-compose` in a `task` for integration testing is effectively doing the same thing (just in one mega-container instead of 2+). It's also worth noting that with a native [Kubernetes Runtime], a single "worker" will likely correspond with an entire cluster, rather than a single node in the cluster.
+
+However, it does mean that we can't provide services to tasks running on Windows/Darwin workers - not sure if there's much need for this, though.
+
+## Networking
+
+The way we accomplish intra-worker container-to-container networking depends on the runtime.
+
+### Guardian and containerd
+
+There are a couple of options here:
+
+1. Containers on the same host can communicate via the bridge network (created by a CNI plugin in our [containerd backend], not sure about Guardian...)
+    * Could work with minimal changes to the runtime layer
+    * Need extra architecture to prevent a malicious `task` from scanning the container subnet to interfere with running services (note: it's currently possible for this to happen with any `task` that runs a server, e.g. running `docker-compose` in a `task`, but is easy to prevent with some changes to firewall rules)
+2. With our containerd runtime, we have more flexibility, and have the option of running both processes in the same network namespace
+    * This would allow communication over `localhost`
+    * We'll need to wait until our containerd runtime is stable so we can replace Guardian with it
+    * If a `task` has multiple services, two services cannot use the same ports (even if they are not exposed)
+
+With respect to the second point under option 1, we *can* prevent such `tasks` if this is a concern by adding a Service Manager component to each worker to register/unregister services. This component could create/destroy firewall rules granting specific containers access to others.
+
+![Service Manager overview](./service-manager.png)
+
+### Kubernetes
+
+When we build a [Kubernetes Runtime], we have a couple alternatives here as well that roughly mirror the choices for containerd:
+
+1. Run the service as its own pod
+    * Possible for service and `task` to run on different k8s nodes
+    * I don't know enough about k8s to know whether it's possible to add firewall rules so that only the `task` pod can access the service pod
```

Comment:
Yep, one can use [network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/)

Reply:
Good to know, thanks! Updated the RFC